src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -554,8 +549,7 @@ protected virtual IEnumerable<string> GetDependsOnRequestIds(IEnumerable<string>
                 }
             }
 
-            ODataBatchUtils.ValidateReferenceUri(uri, flattenDependsOnIds,
-                this.outputContext.MessageWriterSettings.BaseUri, batchFormat);
+            ODataBatchUtils.ValidateReferenceUri(uri, convertedDependsOnIds, this.outputContext.MessageWriterSettings.BaseUri);

```

Comment:
>convertedDependsOnIds [](start = 54, length = 21)

I don't think this is the right logic.  I thought that, if the derived class passed in null for dependsOnIds, we were going to use the ids from the payloadUriConverter (which would be be correctly scoped for the odata version).  The ids from the DependsOnIdsTracker will be return only the ids in the changeset (if currently in changeset) otherwise top level ids, which is correct for 4.0, but wrong for 4.01.  I actually don't think we need the dependsOnIdsTracker at all for the ODataMultipartMixedBatchWriter, only for the ODataMultipartMixedBatchReader.

Reply:
There are pros and cons of using payloadUrlConverter. I thought about using it but ended up not doing it because it is undesirable to couple with legacy issue of difference between changeset scope in V4.0 implementation and batch scope in V4.0 protocol. In other words, current payloadUrlConverter is actually implemented incorrectly against the V4.0 protocol.
Furthermore, Json Batch Format 19.1: (for referencing url...)"The id of this request MUST be specified in the dependsOn name/value pair", it is natural extension to validate the referenced Id in url using the dependsOnIds? 

I understand doing this will allow, on the writer side, the referencing of preceding top level request, e.g. request_2's url uses "$request_1", which should not be allowed in V4.0 implementation due to scope issue above but should not be an problem in V4.01. 

All these said, I just don't want to let the scope issue above limit the validation on the writer side, and the referencing urls in both cases above will still be validated / processed correctly on the reader side.

---
In reply to: [160558905](https://github.com/OData/odata.net/pull/1020#discussion_r160558905) [](ancestors = 160558905)

Reply:
For the case of the non-null dependsOnIds (JSON batch and multipartmixed with explicit dependsOnIds) yes, absolutely we should use the (flattened) list of dependsOnIds explicitly passed by the user.  However, for the dependsOnIds==null (multipartmixed implicit) case, we should default to the set of Ids in the payloadUrlConverter.

---
In reply to: [160567393](https://github.com/OData/odata.net/pull/1020#discussion_r160567393) [](ancestors = 160567393,160558905)

Reply:
updated with tests added.

---
In reply to: [160573714](https://github.com/OData/odata.net/pull/1020#discussion_r160573714) [](ancestors = 160573714,160567393,160558905)